policy/policy-data.go

```diff
@@ -0,0 +1,139 @@
+	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
+		labels[i], labels[j] = labels[j], labels[i]
+	}
+	return strings.Join(labels, ".")
+}
+
+// PolicyAuthorityDatabaseImpl enforces policy decisions based on various rule
+// lists
+type PolicyAuthorityDatabaseImpl struct {
+	log   *blog.AuditLogger
+	dbMap *gorp.DbMap
+}
+
+// NewPolicyAuthorityDatabaseImpl constructs a Policy Authority Database (and
+// creates tables if they are non-existent)
+func NewPolicyAuthorityDatabaseImpl(dbMap *gorp.DbMap) (padb *PolicyAuthorityDatabaseImpl, err error) {
+	logger := blog.GetAuditLogger()
+
+	dbMap.AddTableWithName(DomainRule{}, "ruleList").SetKeys(false, "Host")
+
+	err = dbMap.CreateTablesIfNotExists()
+	if err != nil {
+		return
+	}
+
+	padb = &PolicyAuthorityDatabaseImpl{
+		dbMap: dbMap,
+		log:   logger,
+	}
+
+	return padb, nil
+}
+
+// LoadRules loads the whitelist and blacklist into the database in a transaction
+// deleting any previous content
+func (padb *PolicyAuthorityDatabaseImpl) LoadRules(rules []DomainRule) error {
+	tx, err := padb.dbMap.Begin()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM ruleList")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	for _, r := range rules {
+		r.Host = reverseName(r.Host)
+		tx.Insert(&r)
+	}
+
+	err = tx.Commit()
+	return err
+}
+
+// DumpRules retrieves all DomainRules in the database so they can be written to
+// disk
+func (padb *PolicyAuthorityDatabaseImpl) DumpRules() ([]DomainRule, error) {
+	var dR []DomainRule
+	_, err := padb.dbMap.Select(&dR, "SELECT * FROM ruleList")
+	for _, r := range dR {
+		r.Host = reverseName(r.Host)
+	}
+
+	return dR, err
+}
+
+// CheckRules will query the database for white/blacklist rules that match host,
+// if both whitelist and blacklist rules are found the whitelist will always win
+func (padb *PolicyAuthorityDatabaseImpl) CheckRules(host string, requireWhitelisted bool) error {
+	host = reverseName(host)
+	var rules []DomainRule
+	// Because of how rules are sorted if there is a relevant whitelist AND blacklist
+	// rule we will catch them both, this query will return a maximum of two rules
+	_, err := padb.dbMap.Select(
+		&rules,
+		`SELECT * FROM ruleList WHERE :host >= host ORDER BY host DESC LIMIT 2`,
+		map[string]interface{}{"host": host},
+	)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	var wRules []string
+	var bRules []string
+	for _, r := range rules {
+		switch r.Type {
+		case blacklisted:
+			if strings.HasPrefix(host, r.Host+".") || host == r.Host {
+				bRules = append(bRules, r.Host)
+			}
+		case whitelisted:
+			if host == r.Host {
+				wRules = append(wRules, r.Host)
+			}
+		}
+	}
+
+	if requireWhitelisted && len(wRules) == 0 {
+		return fmt.Errorf("Domain name is not whitelisted for issuance")
+	} else if len(wRules)+len(bRules) > 0 {
+		padb.log.Info(fmt.Sprintf("Hostname [%s] matches rules, Whitelist: %s, Blacklist: %s", host, wRules, bRules))
+		if len(wRules) > 0 {
```

Comment:
Just noting here that on Tuesday's group call, I think we wound up concluding that the blacklist takes precedence, and if we need to whitelist a domain that is covered by the blacklist, we'll also remove it from the blacklist.

This avoids having a weird situation where someone's domain works because they were on the whitelist, but then when we disable whitelisting for GA, their domain stops working because they were also on the blacklist.